src/openzeppelin/security/pausable.cairo

```diff
@@ -1,39 +1,60 @@
-#[contract]
+#[starknet::interface]
+trait IPausable<TState> {
+    fn is_paused(self: @TState) -> bool;
+}
+
+#[starknet::contract]
 mod Pausable {
     use starknet::ContractAddress;
     use starknet::get_caller_address;
 
+    #[storage]
     struct Storage {
         paused: bool
     }
 
     #[event]
-    fn Paused(account: ContractAddress) {}
-
-    #[event]
-    fn Unpaused(account: ContractAddress) {}
-
-    fn is_paused() -> bool {
-        paused::read()
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        Paused: Paused,
+        Unpaused: Unpaused,
     }
-
-    fn assert_not_paused() {
-        assert(!is_paused(), 'Pausable: paused');
+    #[derive(Drop, starknet::Event)]
+    struct Paused {
+        account: ContractAddress
     }
-
-    fn assert_paused() {
-        assert(is_paused(), 'Pausable: not paused');
+    #[derive(Drop, starknet::Event)]
+    struct Unpaused {
+        account: ContractAddress
     }
 
-    fn pause() {
-        assert_not_paused();
-        paused::write(true);
-        Paused(get_caller_address());
+    #[external(v0)]
+    impl PausableImpl of super::IPausable<ContractState> {
+        fn is_paused(self: @ContractState) -> bool {
+            self.paused.read()
+        }
     }
 
-    fn unpause() {
-        assert_paused();
-        paused::write(false);
-        Unpaused(get_caller_address());
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        fn assert_not_paused(self: @ContractState) {
+            assert(!self.paused.read(), 'Pausable: paused');
+        }
+
+        fn assert_paused(self: @ContractState) {
+            assert(self.paused.read(), 'Pausable: not paused');
+        }
+
+        fn _pause(ref self: ContractState) {
+            self.assert_not_paused();
+            self.paused.write(true);
+            self.emit(Event::Paused(Paused { account: get_caller_address() }));
```

Comment:
```suggestion
            self.emit(Paused { account: get_caller_address() });
```

I think this should work as well, right?

Reply:
Good call! This is much easier